File: algorithms/cpp/98.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    bool isValidBST(TreeNode* root)
    {
        if ( !root )
            return true;
        if ( !isValidBST(root->left) || !isValidBST(root->right) )
            return false;
        if ( root->left )
        {
            TreeNode *max_left = root->left;
            while ( max_left->right )
                max_left = max_left->right;
            if ( max_left->val >= root->val )
                return false;
        }
        if ( root->right )
        {
            TreeNode *min_right = root->right;
            while ( min_right->left )
                min_right = min_right->left;
            if ( min_right->val <= root->val )
                return false;
        }
        return true;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        if root.left:
            l = root.left
            while l.right:
                l = l.right
            if l.val >= root.val:
                return False

        if root.right:
            r = root.right
            while r.left:
                r = r.left
            if r.val <= root.val:
                return False

        return self.isValidBST(root.left) and self.isValidBST(root.right)
